llmfoundry/callbacks/kill_loss_spike_callback.py

```diff
@@ -0,0 +1,69 @@
+# Copyright 2022 MosaicML LLM Foundry authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""Track training runs for loss spikes or persistently high training loss."""
+from __future__ import annotations
+
+import torch
+import logging
+import numpy as np
+from composer.core import Callback, State
+from composer.loggers import Logger, MosaicMLLogger
+# from llmfoundry.utils.exceptions import LossSpikeError
+log = logging.getLogger(__name__)
+
+__all__ = ['KillLossSpike']
+
+class KillLossSpike(Callback):
+	
+    def __init__(self, patience:int=4, outlier_multiplier:int=2, window_size:int=100, loss_cap:int=10):
+        self.patience = patience
+        self.outlier_multiplier = outlier_multiplier
+        self.window_size = window_size
+        self.loss_cap = loss_cap
+        self.outlier_counter = 0
+        self.loss_window = []
+
+    def batch_end(self, state: State, logger: Logger) -> None:
+
+        if not isinstance(state.loss, torch.Tensor):
+            raise NotImplementedError('Multiple losses not supported yet')
+        train_loss = state.loss.item()
+
+        # Only start early stopping once a full window of loss data
```

Comment:
can you please factor the detection out into a function that you can then unit test?

Reply:
Done!